drivers/crypto/crypto_intel_sha.c

```diff
@@ -0,0 +1,335 @@
+		if (ret != 0) {
+			return ret;
+		}
+
+		/* set state for next iteration */
+		session->state_.part.hrsm = SHA_HRSM_ENABLE;
+		session->state_.part.state = SHA_MIDLE;
+
+		pkt->in_len -= frag_length;
+		pkt->in_buf += frag_length;
+	} while (pkt->in_len > 0);
+
+	if (finish) {
+		volatile size_t output_size;
+		struct sha_container *self =
+			(struct sha_container *const)(ctx->device)->data;
+		switch (self->dfsha->shactl.part.algo) {
+		case CRYPTO_HASH_ALGO_SHA224:
+			output_size = SHA224_ALGORITHM_HASH_SIZEOF;
+			break;
+		case CRYPTO_HASH_ALGO_SHA256:
+			output_size = SHA256_ALGORITHM_HASH_SIZEOF;
+			break;
+		case CRYPTO_HASH_ALGO_SHA384:
+			output_size = SHA384_ALGORITHM_HASH_SIZEOF;
+			break;
+		case CRYPTO_HASH_ALGO_SHA512:
+			output_size = SHA512_ALGORITHM_HASH_SIZEOF;
+			break;
+		default:
+			return -ENOTSUP;
+		}
+		ret = intel_sha_device_get_hash(ctx->device, pkt->out_buf, output_size);
+
+		/* Fix byte ordering to match common hash representation. */
+		uint32_t *hash_int_ptr = (uint32_t *)(pkt->out_buf);
+
+		for (size_t i = 0; i != output_size / sizeof(uint32_t); i++) {
+			hash_int_ptr[i] = BYTE_SWAP32(hash_int_ptr[i]);
+		}
+	}
+	return ret;
+}
+
+static int intel_sha_device_set_hash_type(struct device *dev, struct hash_ctx *ctx,
+					  enum hash_algo algo)
+{
+	int ret;
+	int ctx_idx;
+	struct sha_container *self = (struct sha_container *const)(dev)->data;
+
+	ctx_idx = intel_sha_get_unused_session_idx();
+
+	if (ctx_idx < 0) {
+		LOG_ERR("All sessions in use!");
+		return -ENOSPC;
+	}
+	ctx->drv_sessn_state = &sha_sessions[ctx_idx];
+
+	/* set processing element enable */
+	self->dfsha->pibcs.part.peen = 0;
+
+	/* populate sha session data */
+	sha_sessions[ctx_idx].state_.part.state = SHA_FIRST;
+	sha_sessions[ctx_idx].state_.part.hrsm = SHA_HRSM_DISABLE;
+	sha_sessions[ctx_idx].algo = algo;
+
+	ctx->hash_hndlr = intel_sha_compute;
+	return ret;
+}
+
+static int intel_sha_device_free(const struct device *dev, struct hash_ctx *ctx)
+{
+	struct sha_container *self = (struct sha_container *const)(dev)->data;
+	struct sha_session *session = (struct sha_session *)ctx->drv_sessn_state;
+
+	(void)memset(self->dfsha, 0, sizeof(struct sha_hw_regs));
+	(void)memset(&session->sha_ctx, 0, sizeof(struct sha_context));
+	(void)memset(&session->state_, 0, sizeof(union sha_state));
+	session->in_use = 0;
+	session->algo = 0;
+	return 0;
+}
+
+static int intel_sha_device_init(struct device *dev)
+{
+	return 0;
+}
+
+static int intel_sha_device_hw_caps(const struct device *dev)
+{
+	return (CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
+}
+
+static struct sha_container sha_data = { .dfsha = SHA_BASE };
+
+static struct crypto_driver_api hash_enc_funcs = {
+	.hash_begin_session = intel_sha_device_set_hash_type,
+	.hash_free_session = intel_sha_device_free,
+	.hash_async_callback_set = NULL,
+	.query_hw_caps = intel_sha_device_hw_caps,
+};
+
+DEVICE_DT_INST_DEFINE(0, intel_sha_device_init, NULL, &sha_data, NULL, POST_KERNEL,
```

Comment:
Is the device limited to 1 instance in the system?

Reply:
Only one instance exists on known hw.